File: src/main/java/buttondevteam/discordplugin/DiscordConnectedPlayer.java
package buttondevteam.discordplugin;

import buttondevteam.discordplugin.mcchat.MinecraftChatModule;
import buttondevteam.discordplugin.playerfaker.DiscordInventory;
import buttondevteam.discordplugin.playerfaker.VCMDWrapper;
import discord4j.core.object.entity.User;
import discord4j.core.object.entity.channel.MessageChannel;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Delegate;
import net.md_5.bungee.api.ChatMessageType;
import net.md_5.bungee.api.chat.BaseComponent;
import org.bukkit.*;
import org.bukkit.attribute.Attribute;
import org.bukkit.attribute.AttributeInstance;
import org.bukkit.attribute.AttributeModifier;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.event.player.PlayerTeleportEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.PlayerInventory;
import org.bukkit.permissions.PermissibleBase;
import org.bukkit.permissions.ServerOperator;
import org.mockito.MockSettings;
import org.mockito.Mockito;

import java.lang.reflect.Modifier;
import java.net.InetSocketAddress;
import java.util.*;

import static org.mockito.Answers.RETURNS_DEFAULTS;

public abstract class DiscordConnectedPlayer extends DiscordSenderBase implements IMCPlayer<DiscordConnectedPlayer> {
	private @Getter VCMDWrapper vanillaCmdListener;
	@Getter
	@Setter
	private boolean loggedIn = false;

	@Delegate(excludes = ServerOperator.class)
	private PermissibleBase origPerm;

	private @Getter String name;

	private @Getter OfflinePlayer basePlayer;

	@Getter
	@Setter
	private PermissibleBase perm;

	private Location location;

	private final MinecraftChatModule module;

	@Getter
	private final UUID uniqueId;

	/**
	 * The parameters must match with {@link #create(User, MessageChannel, UUID, String, MinecraftChatModule)}
	 */
	protected DiscordConnectedPlayer(User user, MessageChannel channel, UUID uuid, String mcname,
	                                 MinecraftChatModule module) {
		super(user, channel);
		location = Bukkit.getWorlds().get(0).getSpawnLocation();
		origPerm = perm = new PermissibleBase(basePlayer = Bukkit.getOfflinePlayer(uuid));
		name = mcname;
		this.module = module;
		uniqueId = uuid;
		displayName = mcname;
		vanillaCmdListener = new VCMDWrapper(VCMDWrapper.createListener(this, module));
	}

	/**
	 * For testing
	 */
	protected DiscordConnectedPlayer(User user, MessageChannel channel) {
		super(user, channel);
		module = null;
		uniqueId = UUID.randomUUID();
	}

	public void setOp(boolean value) { //CraftPlayer-compatible implementation
		this.origPerm.setOp(value);
		this.perm.recalculatePermissions();
	}

	public boolean isOp() { return this.origPerm.isOp(); }

	@Override
	public boolean teleport(Location location) {
		if (module.allowFakePlayerTeleports.get())
			this.location = location;
		return true;
	}

	@Override
	public boolean teleport(Location location, PlayerTeleportEvent.TeleportCause cause) {
		if (module.allowFakePlayerTeleports.get())
			this.location = location;
		return true;
	}

	@Override
	public boolean teleport(Entity destination) {
		if (module.allowFakePlayerTeleports.get())
			this.location = destination.getLocation();
		return true;
	}

	@Override
	public boolean teleport(Entity destination, PlayerTeleportEvent.TeleportCause cause) {
		if (module.allowFakePlayerTeleports.get())
			this.location = destination.getLocation();
		return true;
	}

	@Override
	public Location getLocation(Location loc) {
		if (loc != null) {
			loc.setWorld(getWorld());
			loc.setX(location.getX());
			loc.setY(location.getY());
			loc.setZ(location.getZ());
			loc.setYaw(location.getYaw());
			loc.setPitch(location.getPitch());
		}

		return loc;
	}

	@Override
	public Server getServer() {
		return Bukkit.getServer();
	}

	@Override
	public void sendRawMessage(String message) {
		sendMessage(message);
	}

	@Override
	public void chat(String msg) {
		Bukkit.getPluginManager()
			.callEvent(new AsyncPlayerChatEvent(true, this, msg, new HashSet<>(Bukkit.getOnlinePlayers())));
	}

	@Override
	public World getWorld() {
		return Bukkit.getWorlds().get(0);
	}

	@Override
	public boolean isOnline() {
		return true;
	}

	@Override
	public Location getLocation() {
		return new Location(getWorld(), location.getX(), location.getY(), location.getZ(),
			location.getYaw(), location.getPitch());
	}

	@Override
	public Location getEyeLocation() {
		return getLocation();
	}

	@Override
	@Deprecated
	public double getMaxHealth() {
		return 20;
	}

	@Override
	public Player getPlayer() {
		return this;
	}

	@Getter
	@Setter
	private String displayName;

	@Override
	public AttributeInstance getAttribute(Attribute attribute) {
		return new AttributeInstance() {
			@Override
			public Attribute getAttribute() {
				return attribute;
			}

			@Override
			public double getBaseValue() {
				return getDefaultValue();
			}

			@Override
			public void setBaseValue(double value) {
			}

			@Override
			public Collection<AttributeModifier> getModifiers() {
				return Collections.emptyList();
			}

			@Override
			public void addModifier(AttributeModifier modifier) {
			}

			@Override
			public void removeModifier(AttributeModifier modifier) {
			}

			@Override
			public double getValue() {
				return getDefaultValue();
			}

			@Override
			public double getDefaultValue() {
				return 20; //Works for max health, should be okay for the rest
			}
		};
	}

	@Override
	public GameMode getGameMode() {
		return GameMode.SPECTATOR;
	}

	@SuppressWarnings("deprecation")
	private final Player.Spigot spigot = new Player.Spigot() {
		@Override
		public InetSocketAddress getRawAddress() {
			return null;
		}

		@Override
		public void playEffect(Location location, Effect effect, int id, int data, float offsetX, float offsetY, float offsetZ, float speed, int particleCount, int radius) {
		}

		@Override
		public boolean getCollidesWithEntities() {
			return false;
		}

		@Override
		public void setCollidesWithEntities(boolean collides) {
		}

		@Override
		public void respawn() {
		}

		@Override
		public String getLocale() {
			return "en_us";
		}

		@Override
		public Set<Player> getHiddenPlayers() {
			return Collections.emptySet();
		}

		@Override
		public void sendMessage(BaseComponent component) {
			DiscordConnectedPlayer.super.sendMessage(component.toPlainText());
		}

		@Override
		public void sendMessage(BaseComponent... components) {
			for (var component : components)
				sendMessage(component);
		}

		@Override
		public void sendMessage(ChatMessageType position, BaseComponent component) {
			sendMessage(component); //Ignore position
		}

		@Override
		public void sendMessage(ChatMessageType position, BaseComponent... components) {
			sendMessage(components); //Ignore position
		}

		@Override
		public boolean isInvulnerable() {
			return true;
		}
	};

	@Override
	public Player.Spigot spigot() {
		return spigot;
	}

	public static DiscordConnectedPlayer create(User user, MessageChannel channel, UUID uuid, String mcname,
	                                            MinecraftChatModule module) {
		return Mockito.mock(DiscordConnectedPlayer.class,
			getSettings().useConstructor(user, channel, uuid, mcname, module));
	}

	public static DiscordConnectedPlayer createTest() {
		return Mockito.mock(DiscordConnectedPlayer.class, getSettings().useConstructor(null, null));
	}

	private static MockSettings getSettings() {
		return Mockito.withSettings()
			.defaultAnswer(invocation -> {
				try {
					if (!Modifier.isAbstract(invocation.getMethod().getModifiers()))
						return invocation.callRealMethod();
					if (PlayerInventory.class.isAssignableFrom(invocation.getMethod().getReturnType()))
						return Mockito.mock(DiscordInventory.class, Mockito.withSettings().extraInterfaces(PlayerInventory.class));
					if (Inventory.class.isAssignableFrom(invocation.getMethod().getReturnType()))
						return new DiscordInventory();
					return RETURNS_DEFAULTS.answer(invocation);
				} catch (Exception e) {
					System.err.println("Error in mocked player!");
					e.printStackTrace();
					return RETURNS_DEFAULTS.answer(invocation);
				}
			})
			.stubOnly();
	}
}

File: src/main/java/buttondevteam/discordplugin/DiscordConnectedPlayer.scala
package buttondevteam.discordplugin

import buttondevteam.discordplugin.mcchat.MinecraftChatModule
import buttondevteam.discordplugin.playerfaker.{DiscordInventory, VCMDWrapper}
import discord4j.core.`object`.entity.User
import discord4j.core.`object`.entity.channel.MessageChannel
import net.md_5.bungee.api.ChatMessageType
import net.md_5.bungee.api.chat.BaseComponent
import org.bukkit._
import org.bukkit.attribute.{Attribute, AttributeInstance, AttributeModifier}
import org.bukkit.entity.Player.Spigot
import org.bukkit.entity.{Entity, Player}
import org.bukkit.event.player.{AsyncPlayerChatEvent, PlayerTeleportEvent}
import org.bukkit.inventory.{Inventory, PlayerInventory}
import org.bukkit.permissions.{PermissibleBase, Permission, PermissionAttachment, PermissionAttachmentInfo}
import org.bukkit.plugin.Plugin
import org.mockito.Answers.RETURNS_DEFAULTS
import org.mockito.{MockSettings, Mockito}
import org.mockito.invocation.InvocationOnMock

import java.lang.reflect.Modifier
import java.net.InetSocketAddress
import java.util
import java.util._

object DiscordConnectedPlayer {
    def create(user: User, channel: MessageChannel, uuid: UUID, mcname: String, module: MinecraftChatModule): DiscordConnectedPlayer =
        Mockito.mock(classOf[DiscordConnectedPlayer], getSettings.useConstructor(user, channel, uuid, mcname, module))

    def createTest: DiscordConnectedPlayer =
        Mockito.mock(classOf[DiscordConnectedPlayer], getSettings.useConstructor(null, null))

    private def getSettings: MockSettings = Mockito.withSettings.defaultAnswer((invocation: InvocationOnMock) => {
        def foo(invocation: InvocationOnMock): AnyRef =
            try {
                if (!Modifier.isAbstract(invocation.getMethod.getModifiers))
                    invocation.callRealMethod
                else if (classOf[PlayerInventory].isAssignableFrom(invocation.getMethod.getReturnType))
                    Mockito.mock(classOf[DiscordInventory], Mockito.withSettings.extraInterfaces(classOf[PlayerInventory]))
                else if (classOf[Inventory].isAssignableFrom(invocation.getMethod.getReturnType))
                    new DiscordInventory
                else
                    RETURNS_DEFAULTS.answer(invocation)
            } catch {
                case e: Exception =>
                    System.err.println("Error in mocked player!")
                    e.printStackTrace()
                    RETURNS_DEFAULTS.answer(invocation)
            }

        foo(invocation)
    }).stubOnly
}

abstract class DiscordConnectedPlayer(user: User, channel: MessageChannel) extends DiscordSenderBase(user, channel) with IMCPlayer[DiscordConnectedPlayer] {
    override def isPermissionSet(name: String): Boolean = this.origPerm.isPermissionSet(name)

    override def isPermissionSet(perm: Permission): Boolean = this.origPerm.isPermissionSet(perm)

    override def hasPermission(inName: String): Boolean = this.origPerm.hasPermission(inName)

    override def hasPermission(perm: Permission): Boolean = this.origPerm.hasPermission(perm)

    override def addAttachment(plugin: Plugin, name: String, value: Boolean): PermissionAttachment = this.origPerm.addAttachment(plugin, name, value)

    override def addAttachment(plugin: Plugin): PermissionAttachment = this.origPerm.addAttachment(plugin)

    override def removeAttachment(attachment: PermissionAttachment): Unit = this.origPerm.removeAttachment(attachment)

    override def recalculatePermissions(): Unit = this.origPerm.recalculatePermissions()

    def clearPermissions(): Unit = this.origPerm.clearPermissions()

    override def addAttachment(plugin: Plugin, name: String, value: Boolean, ticks: Int): PermissionAttachment =
        this.origPerm.addAttachment(plugin, name, value, ticks)

    override def addAttachment(plugin: Plugin, ticks: Int): PermissionAttachment = this.origPerm.addAttachment(plugin, ticks)

    override def getEffectivePermissions: util.Set[PermissionAttachmentInfo] = this.origPerm.getEffectivePermissions

    def setLoggedIn(loggedIn: Boolean): Unit = this.loggedIn = loggedIn

    def setPerm(perm: PermissibleBase): Unit = this.perm = perm

    override def setDisplayName(displayName: String): Unit = this.displayName = displayName

    override def getVanillaCmdListener: VCMDWrapper = this.vanillaCmdListener

    def isLoggedIn: Boolean = this.loggedIn

    override def getName: String = this.name

    def getBasePlayer: OfflinePlayer = this.basePlayer

    def getPerm: PermissibleBase = this.perm

    override def getUniqueId: UUID = this.uniqueId

    override def getDisplayName: String = this.displayName

    private var vanillaCmdListener: VCMDWrapper = null
    private var loggedIn = false
    private var origPerm: PermissibleBase = null
    private var name: String = null
    private var basePlayer: OfflinePlayer = null
    private var perm: PermissibleBase = null
    private var location: Location = null
    final private var module: MinecraftChatModule = null
    final private var uniqueId: UUID = null
    final private var displayName: String = null

    /**
     * The parameters must match with {@link #create ( User, MessageChannel, UUID, String, MinecraftChatModule)}
     */
    def this(user: User, channel: MessageChannel, uuid: UUID, mcname: String, module: MinecraftChatModule) {
        this(user, channel)
        location = Bukkit.getWorlds.get(0).getSpawnLocation
        perm = new PermissibleBase(basePlayer = Bukkit.getOfflinePlayer(uuid))
        origPerm = perm
        name = mcname
        this.module = module
        uniqueId = uuid
        displayName = mcname
        vanillaCmdListener = new VCMDWrapper(VCMDWrapper.createListener(this, module))
    }

    /**
     * For testing
     */
    def this(user: User, channel: MessageChannel) {
        this(user, channel)
        module = null
        uniqueId = UUID.randomUUID
    }

    override def setOp(value: Boolean): Unit = { //CraftPlayer-compatible implementation
        this.origPerm.setOp(value)
        this.perm.recalculatePermissions()
    }

    override def isOp: Boolean = this.origPerm.isOp

    override def teleport(location: Location): Boolean = {
        if (module.allowFakePlayerTeleports.get) this.location = location
        true
    }

    def teleport(location: Location, cause: PlayerTeleportEvent.TeleportCause): Boolean = {
        if (module.allowFakePlayerTeleports.get) this.location = location
        true
    }

    override def teleport(destination: Entity): Boolean = {
        if (module.allowFakePlayerTeleports.get) this.location = destination.getLocation
        true
    }

    def teleport(destination: Entity, cause: PlayerTeleportEvent.TeleportCause): Boolean = {
        if (module.allowFakePlayerTeleports.get) this.location = destination.getLocation
        true
    }

    override def getLocation(loc: Location): Location = {
        if (loc != null) {
            loc.setWorld(getWorld)
            loc.setX(location.getX)
            loc.setY(location.getY)
            loc.setZ(location.getZ)
            loc.setYaw(location.getYaw)
            loc.setPitch(location.getPitch)
        }
        loc
    }

    override def getServer: Server = Bukkit.getServer

    override def sendRawMessage(message: String): Unit = sendMessage(message)

    override def chat(msg: String): Unit = Bukkit.getPluginManager.callEvent(new AsyncPlayerChatEvent(true, this, msg, new util.HashSet[Player](Bukkit.getOnlinePlayers)))

    override def getWorld: World = Bukkit.getWorlds.get(0)

    override def isOnline = true

    override def getLocation = new Location(getWorld, location.getX, location.getY, location.getZ, location.getYaw, location.getPitch)

    override def getEyeLocation: Location = getLocation

    @deprecated override def getMaxHealth = 20

    override def getPlayer: DiscordConnectedPlayer = this

    override def getAttribute(attribute: Attribute): AttributeInstance = new AttributeInstance() {
        override def getAttribute: Attribute = attribute

        override def getBaseValue: Double = getDefaultValue

        override def setBaseValue(value: Double): Unit = {
        }

        override def getModifiers: util.Collection[AttributeModifier] = Collections.emptyList

        override def addModifier(modifier: AttributeModifier): Unit = {
        }

        override def removeModifier(modifier: AttributeModifier): Unit = {
        }

        override def getValue: Double = getDefaultValue

        override def getDefaultValue: Double = 20 //Works for max health, should be okay for the rest
    }

    override def getGameMode = GameMode.SPECTATOR

    //noinspection ScalaDeprecation
    @SuppressWarnings(Array("deprecation")) final private val spigot: Spigot = new Spigot() {
        override def getRawAddress: InetSocketAddress = null

        override def playEffect(location: Location, effect: Effect, id: Int, data: Int, offsetX: Float, offsetY: Float, offsetZ: Float, speed: Float, particleCount: Int, radius: Int): Unit = {
        }

        override def getCollidesWithEntities = false

        override def setCollidesWithEntities(collides: Boolean): Unit = {
        }

        override def respawn(): Unit = {
        }

        override def getLocale = "en_us"

        override def getHiddenPlayers: util.Set[Player] = Collections.emptySet

        override def sendMessage(component: BaseComponent): Unit =
            DiscordConnectedPlayer.super.sendMessage(component.toPlainText)

        override def sendMessage(components: BaseComponent*): Unit =
            for (component <- components)
                sendMessage(component)

        override def sendMessage(position: ChatMessageType, component: BaseComponent): Unit =
            sendMessage(component) //Ignore position
        override def sendMessage(position: ChatMessageType, components: BaseComponent*) =
            sendMessage(components)

        override def isInvulnerable = true
    }

    override def spigot: Spigot = spigot
}